### DISCUSSION TITLE
Thank You / Import Mechanism

### DISCUSSION THREAD

**Discussion by Mykbibby** [neanes/neanes#290] [Ideas]
Hey there,

First off -- I just want to say this program is a literal godsend. I am not a coder, but have been dreaming of this app for so many years. I prayed that there would be some chanter out there like me who would see the massive void in the market for this app so that I wouldn't have to get the horrific Melodos with a windows machine just devoted to its use. And then you came around and designed this literally perfect implementation of the dream. Thank. You.

My one single feature request would be some way for an ordinary PDF to be imported into the program, and for Neanes to do its best at creating a starting point for the score. Even if there are errors, these could be corrected manually. Perhaps this is way to difficult of a technical ask, but I figured I'd at least put it out there.

Thanks again!

**Comment by danielgarthur** [comment]
I would like to include OCR in Neanes, but I don't have much experience in this area. I do have some experience with OpenCV and TensorFlow, but in the context of detecting object in live video. However, I wonder if this could work for Byzantine neumes, too. It might be time consuming to create and train the model, though, and I have doubts that handwritten scores would ever work very well. But importing scores from classical books seems like it should be possible.

**Comment by danielgarthur** [comment]
A small update:

I've been working on a proof-of-concept app in C++ that uses OpenCV and the k-NN algorithm to detect neumes. I'll eventually post the code in a public repository once it's in a more stable state.

What I have so far is a native app that trains a k-NN model from a directory structure of images and then, given an image of a page from a book, it

- Detects neume base lines
- Detects text lines
- Removes text
- Recognizes neumes
- Classifies neumes into groups (i.e. associates supporting neumes with the correct base neumes) and sorts them
- Outputs results as a YAML file

A secondary script converts the YAML to a Neanes file.

**Reply by Mykbibby** [reply]
This is sensational news! I'm unbelievably excited for this. However flawed it may be, it will still speed up the process immensely compared to starting from scratch, especially when working with books. Thanks so much for everything you do Daniel!

**Comment by basil** [comment]
That's pretty cool. Though I worry that a large influx of users importing data into this file format is going to put pressure on us to stabilize the file format. Right now I think the object model still has some deficiencies, mostly having to do with sharps and flats.

**Reply by danielgarthur** [reply]
I intend this to be a separate C++ library that anyone can interface using its data model, which is pretty low level. The burden of converting to Neanes's data model will remain with Neanes.

The output from the library does very little interpretation. It just says "I saw a neume group with an oligon base with two kentimata and a gorgon as supporting neumes. Here are their coordinates". 

It's Neanes' job to look at that and check the coordinates of the kentimata whether they are above or below the oligon, etc.
